In [1]:
import os
import re
import json
import tqdm
import numpy as np
import ray
from transformers import GPT2Tokenizer
import openai
import ast
import time

In [2]:
class RedundantParenthesesRemover(ast.NodeTransformer):
    def visit_Expr(self, node):
        self.generic_visit(node)
        if isinstance(node.value, ast.BinOp):
            return node.value
        return node

def remove_redundant_parentheses(expression):
    # Parse the expression
    parsed_expression = ast.parse(expression)

    # Remove redundant parentheses
    transformer = RedundantParenthesesRemover()
    transformed_expression = transformer.visit(parsed_expression)

    # Convert the transformed expression back to a string
    simplified_expression = ast.unparse(transformed_expression)

    return simplified_expression
    
def divide(a, b):
    return f"({a} / {b})"

def subtract(a, b):
    return f"({a} - {b})"

def multiply(a, b):
    return f"({a} * {b})"

def add(a, b):
    return f"({a} + {b})"

def exp(a, b):
    return f"({a} ** {b})"

def greater(a, b):
    return f"({a} > {b})"

def translate_expr(expr):
    if "table" in expr:
        return expr
    
    # replace const_m1
    expr = re.sub(r'const_m1', r'-1', expr)

    # change % to / 100
    expr = re.sub(r'([0-9]*\.?[0-9]+)%', r'divide(\1 , 100)', expr)
    
    expr = re.sub(r'const_([0-9]*\.?[0-9]+)', r'\1', expr)
    try:
        new_expr = eval(expr)
        new_expr = remove_redundant_parentheses(new_expr)
    except Exception as e:
        print(e, expr)
        new_expr = expr
    
    return new_expr

def convert_to_markdown(data):
    markdown = "|"
    
    # Add table headers
    for header in data[0]:
        markdown += header + "|"
    markdown += "\n|"
    
    # Add table header separators
    for _ in data[0]:
        markdown += "---|"
    markdown += "\n"
    
    # Add table rows
    for row in data[1:]:
        markdown += "|"
        for cell in row:
            markdown += cell + "|"
        markdown += "\n"
        
    return markdown

def extract_answer(response):
    # extract content inside Calculate()
    matches = re.findall(r"Calculate\(([\(\)0-9 ><,\.\/\+\-\*]*)\)", response)
    if len(matches) == 0:
        if "Yes" in response:
            return "Yes"
        elif "No" in response:
            return "No"
        else:
            return ""
    else:
        output = matches[0].replace(",", "")
        return output
    
def if_exec_correct(t_prog, g_prog):
    try:
        t_exec = eval(t_prog)

        if type(t_exec) == bool and g_prog in ["Yes", "No"]:
            t_exec = "Yes" if t_exec else "No"

            if t_exec == g_prog:
                return True
            
        g_exec = eval(g_prog)

        if t_exec == g_exec:
            return True
        elif t_exec * 100 == g_exec:
            return True
        elif t_exec * 100 == -g_exec:
            return True
        elif t_exec == g_exec * 100:
            return True
        elif t_exec == -g_exec * 100:
            return True
        elif t_exec * 1000000 == g_exec:
            return True
        elif t_exec * 1000000 == -g_exec:
            return True
        elif t_exec == g_exec * 1000000:
            return True
        elif t_exec == -g_exec * 1000000:
            return True
        elif t_exec == -g_exec:
            return True
    except:
        return False

    return False

In [3]:
system_prompt = (
                "You need to answer the user's question in the ### Question ### section.\n" \
                "You need to provide the answer in the format 'Calculate(a + b)', where the expression needs to be python excutable." \
                # "You can calculate the average of a column by using the function 'Average(table_column_name)'.\n" \
                # "Similarly, you can calculate the sum, the maximum, the minimum, the count of a column by using the functions "\
                # "'Sum(table_column_name)', 'Max(table_column_name)', 'Min(table_column_name)', 'Count(table_column_name)' respectively.\n" \
                # "You only use the table's column name inside those operations\n" \
                "For example, if the question is 'What is the sum of 1 + 2?', you need to answer 'Calc(1 + 2)'." \
                "if the question is 'Is 123 greater than 231?', you need to answer 'Calc(123 > 231)'." \
                # "|Age|\n|---|\n|12|\n|15|\n|16|\n\n What is the average age? The answer is 'Calculate(Average(Age))'" \
                "DO NOT give anything else other than'Calculate()'."
                )

In [4]:
user_example1 = '### Context ###\n\ndebt maturities 2013 the following table presents aggregate debt maturities as of december 31 , 2007 , excluding market value adjustments .\nmillions of dollars .\n\n|2008|$ 689|\n|---|---|\n|2009|542|\n|2010|462|\n|2011|550|\n|2012|720|\n|Thereafter|4,717|\n|Total debt|$ 7,680|\n\n\nat december 31 , 2007 , we reclassified as long-term debt approximately $ 550 million of debt due within one year that we intend to refinance .\nthis reclassification reflected our ability and intent to refinance any short- term borrowings and certain current maturities of long-term debt on a long-term basis .\nat december 31 , 2006 , we did not reclassify any short-term debt as long-term debt as we did not intend to refinance at that mortgaged properties 2013 equipment with a carrying value of approximately $ 2.8 billion at both december 31 , 2007 and 2006 serves as collateral for capital leases and other types of equipment obligations in accordance with the secured financing arrangements utilized to acquire such railroad equipment .\nas a result of the merger of missouri pacific railroad company ( mprr ) with and into uprr on january 1 , 1997 , and pursuant to the underlying indentures for the mprr mortgage bonds , uprr must maintain the same value of assets after the merger in order to comply with the security requirements of the mortgage bonds .\nas of the merger date , the value of the mprr assets that secured the mortgage bonds was approximately $ 6.0 billion .\nin accordance with the terms of the indentures , this collateral value must be maintained during the entire term of the mortgage bonds irrespective of the outstanding balance of such bonds .\ncredit facilities 2013 on december 31 , 2007 , $ 1.9 billion of credit was available under our revolving credit facility ( the facility ) , which we entered into on april 20 , 2007 .\nthe facility is designated for general corporate purposes and supports the issuance of commercial paper .\nwe did not draw on the facility during 2007 .\ncommitment fees and interest rates payable under the facility are similar to fees and rates available to comparably rated , investment-grade borrowers .\nthe facility allows for borrowings at floating rates based on london interbank offered rates , plus a spread , depending upon our senior unsecured debt ratings .\nthe facility requires the maintenance of a debt to net worth coverage ratio .\nat december 31 , 2007 , we were in compliance with this covenant .\nthe facility does not include any other financial restrictions , credit rating triggers ( other than rating-dependent pricing ) , or any other provision that could require us to post collateral .\nthe facility , which expires in april 2012 , replaced two $ 1 billion , 5-year facilities with terms ending in march 2009 and march 2010 .\nthe facility includes terms that are comparable with those of the prior facilities , although the minimum net worth requirement of $ 7.5 billion in prior facilities was removed , and the facility includes a change-of-control provision .\nin addition to our revolving credit facility , a $ 75 million uncommitted line of credit was available .\nthe line of credit expires in april 2008 , and was not used in 2007 .\nwe must have equivalent credit available under our five-year facility to draw on this $ 75 million line .\ndividend restrictions 2013 our revolving credit facility includes a debt-to-net worth covenant that , under certain circumstances , would restrict the payment of cash dividends to our shareholders .\nthe amount of retained earnings available for dividends was $ 11.5 billion and $ 7.8 billion at december 31 , 2007 and december 31 , 2006 , respectively .\nthis facility replaced two credit facilities that had minimum net worth covenants that were more restrictive with respect to the amount of retained earnings available for dividends at december 31 , 2006. .\n\n### Question ###\n\nwhat percentage of total debt is due after 2012?'
assistant_example1 = "Calculate(4717 / 7680)"

In [7]:
# @ray.remote
def  get_responses(rank, model_name, data):
    openai.api_key = "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"
    responses = []

    if rank == 0:
        data = tqdm.tqdm(data)
    else:
        data = data

    for item in data:
        table_md = convert_to_markdown(item["table_ori"])
        question = item["qa"]["question"]
        
        pre_text = "\n".join(item["pre_text"])
        post_text = "\n".join(item["post_text"])
        
        user_prompt = f"### Context ###\n\n{pre_text}\n\n{table_md}\n\n{post_text}\n\n### Question ###\n\n{question}"

        messages = [{"role": "system", 
                    "content": system_prompt
                    },
                    {"role": "user",
                    "content": user_example1,
                    },
                    {"role": "assistant",
                    "content": assistant_example1
                    }
                    ]

        messages.append({"role": "user", "content": user_prompt})


        success = False
        while not success:
            try:
                completion = openai.ChatCompletion.create(model=model_name, 
                                                        messages=messages,
                                                        top_p=0.9,
                                                        temperature=0.3,)
                success = True
            except Exception as e:
                print(e)
                time.sleep(1)
        
        result = completion.choices[0].message.content
        
        responses.append(result)
    
    return responses

In [6]:
filepath = "../FinQA/dataset/test.json"

with open(filepath) as f:
    data = json.load(f)

programs = []
translated_programs = []
answers = []

for item in tqdm.tqdm(data):
    table_md = convert_to_markdown(item["table_ori"])
    question = item["qa"]["question"]
    
    pre_text = "\n".join(item["pre_text"])
    post_text = "\n".join(item["post_text"])

    programs.append(item["qa"]["program_re"])
    translated_programs.append(translate_expr(item["qa"]["program_re"]))
    answers.append(item["qa"]["answer"])

100%|██████████| 1147/1147 [00:00<00:00, 13594.74it/s]


In [7]:
# responses = get_responses(0, data)

In [21]:
with open("results/gpt-3.5.json") as f:
    responses = json.load(f)
    responses = responses["responses"]

In [22]:
with open("results/gpt-4.json") as f:
    redo_responses = json.load(f)
    # responses = responses["responses"]
    wrong_indices = redo_responses["wrong_indices"]
    redo_responses = redo_responses["responses"]

In [23]:
for i in range(len(redo_responses)):
    responses[wrong_indices[i]] = redo_responses[i]

In [25]:
gen_programs = []

for response in responses:
    gen_programs.append(extract_answer(response))

In [26]:
# calculate accuracy
prog_correct = 0
exec_correct = 0

# wrong_indices = []

for i in range(len(translated_programs)):
    t_prog = translated_programs[i].replace(" ", "")
    g_prog = gen_programs[i].replace(" ", "")

    if t_prog == g_prog:
        prog_correct += 1

    if if_exec_correct(t_prog, g_prog):
        exec_correct += 1
    else:
        # wrong_indices.append(i)
        print(i, t_prog, "|", g_prog)

4 (318.46-100)/100-(206.49-100)/100 | (318.46-206.49)/206.49
5 (92710000-86842000)/86842000 | (100690000-92710000)/92710000
15 table_average(netchangefortheyear,none) | (4648+3625+1620)/3
17 (703.1+705.4+2)/2 | (705.4+703.1)/2
20 (772-843)/843 | 
21 (311.81-100)/100-(198.18-100)/100 | 311.81-198.18
22 (101.88-93.21)/93.21 | (101.88-93.21)/101.88
28 0.6+0.5+4.7 | 0.5+0.6+4.7+0.6+0.2+0.2
29 3500/3081 | 3081/(150+175+2756)
32 table_max(provisionforincometaxes,none) | 
34 61912-367 | 51052-367
35 168-56 | 224-56
36 165000/254000 | (254000-165000)/254000
37 166000/219000 | (219000-166000)/219000
41 (248.36-100)/100-(236.2-100)/100 | (248.36-236.20)/236.20
48 10.4/61.4*100 | 10354/61400
51 2.3/0.5 | 19.3/0.5
59 411636+439000 | (411636+439000+556000)/3
75 11.02/9.52/9.52 | (11.02-9.52)/9.52
79 207.62/97.13 | 207.62-97.13
80 207.62-97.13 | 207.62-107.76
81 37/1.4 | 37*(1-(1/1.4))
84 (1.11+1.08)/1.08 | (1.11-1.08)/1.08
92 table_average(operatingincome,none) | (762.6+808.4+895.2)/3
99 7532/18099

In [27]:
prog_correct / len(translated_programs), exec_correct / len(translated_programs)

(0.3687881429816914, 0.7567567567567568)

In [16]:
exec_correct / len(translated_programs)

0.5745422842197035

In [17]:
with open("gpt-4.json") as f:
    gpt4_responses = json.load(f)
    wrong_indices = gpt4_responses["wrong_indices"]
    gpt4_responses = gpt4_responses["responses"]

In [18]:
for i in range(len(wrong_indices)):
    responses[wrong_indices[i]] = gpt4_responses[i]

In [19]:
gen_programs = []

for response in responses:
    gen_programs.append(extract_answer(response))

In [20]:
# calculate accuracy
prog_correct = 0
exec_correct = 0

for i in range(len(translated_programs)):
    t_prog = translated_programs[i].replace(" ", "")
    g_prog = gen_programs[i].replace(" ", "")

    if t_prog == g_prog:
        prog_correct += 1

    if if_exec_correct(t_prog, g_prog):
        exec_correct += 1
    else:
        print(i, t_prog, "|", g_prog)

4 (318.46-100)/100-(206.49-100)/100 | (318.46-206.49)/206.49
5 (92710000-86842000)/86842000 | (100690000-92710000)/92710000
15 table_average(netchangefortheyear,none) | (4648+3625+1620)/3
17 (703.1+705.4+2)/2 | (705.4+703.1)/2
20 (772-843)/843 | 
21 (311.81-100)/100-(198.18-100)/100 | 311.81-198.18
22 (101.88-93.21)/93.21 | (101.88-93.21)/101.88
28 0.6+0.5+4.7 | 0.5+0.6+4.7+0.6+0.2+0.2
29 3500/3081 | 3081/(150+175+2756)
32 table_max(provisionforincometaxes,none) | 
34 61912-367 | 51052-367
35 168-56 | 224-56
36 165000/254000 | (254000-165000)/254000
37 166000/219000 | (219000-166000)/219000
41 (248.36-100)/100-(236.2-100)/100 | (248.36-236.20)/236.20
48 10.4/61.4*100 | 10354/61400
51 2.3/0.5 | 19.3/0.5
59 411636+439000 | (411636+439000+556000)/3
75 11.02/9.52/9.52 | (11.02-9.52)/9.52
79 207.62/97.13 | 207.62-97.13
80 207.62-97.13 | 207.62-107.76
81 37/1.4 | 37*(1-(1/1.4))
84 (1.11+1.08)/1.08 | (1.11-1.08)/1.08
92 table_average(operatingincome,none) | (762.6+808.4+895.2)/3
99 7532/18099

In [22]:
prog_correct / len(translated_programs)

0.3687881429816914

In [115]:
system_prompt = (
                "You need to answer the user's question in the ### Question ### section.\n" \
                "You need to provide the answer in the format 'Calculate(expr)', where the expression needs to be python excutable." \
                # "You can calculate the average of a column by using the function 'Average(table_column_name)'.\n" \
                # "Similarly, you can calculate the sum, the maximum, the minimum, the count of a column by using the functions "\
                # "'Sum(table_column_name)', 'Max(table_column_name)', 'Min(table_column_name)', 'Count(table_column_name)' respectively.\n" \
                # "You only use the table's column name inside those operations\n" \
                # "For example, if the question is 'What is the sum of 1 + 2?', you need to answer 'Calc(1 + 2)'." \
                # "if the question is 'Is 123 greater than 231?', you need to answer 'Calc(123 > 231)'." \
                # "|Age|\n|---|\n|12|\n|15|\n|16|\n\n What is the average age? The answer is 'Calculate(Average(Age))'" \
                # "DO NOT give anything else other than'Calculate()'."
                )

In [137]:
item["id"]

'AMT/2016/page_49.pdf-1'

In [58]:
item = data[1010]
table_md = convert_to_markdown(item["table_ori"])
question = item["qa"]["question"]
pre_text = "\n".join(item["pre_text"])
post_text = "\n".join(item["post_text"])

user_prompt = f"### Context ###\n\n{pre_text}\n\n{table_md}\n\n{post_text}\n\n### Question ###\n\n{question}"

In [59]:
print(user_prompt)

### Context ###

american tower corporation and subsidiaries notes to consolidated financial statements 2014 ( continued ) to purchase 3924 and 911 shares , respectively .
in october 2005 , in connection with the exercise by mr .
gearon of his right to require the company to purchase his interest in atc south america , these options vested in full and were exercised .
upon exercise of these options , the holders received 4428 shares of atc south america , net of 1596 shares retained by the company to satisfy employee tax withholding obligations .
the 1596 shares retained by the company were treated as a repurchase of a minority interest in accordance with sfas no .
141 .
as a result , the company recorded a purchase price allocation adjustment of $ 5.6 million as an increase to intangible assets and a corresponding increase in minority interest as of the date of acquisition .
the holders had the right to require the company to purchase their shares of atc south america at their then fa

: 

||Liability as of January 1, 2004|2004 Expense|2004 Cash Payments|Liability as of December 31, 2004|2005 Expense|2005 Cash Payments|Liability as of December 31, 2005|2006 Expense|2006 Cash Payments|Liability as of December 31, 2006|
|---|---|---|---|---|---|---|---|---|---|---|
|Employee separations|$2,239|$823|$(2,397)|$665|$84|$(448)|$301|$(267)|$(34)|$0|
|Lease terminations and other facility closing costs|1,450|(131)|(888)|431|12|(325)|118|(10)|(108)|0|
|Total|$3,689|$692|$(3,285)|$1,096|$96|$(773)|$419|$(277)|$(142)|$0|


In [153]:
translated_programs[79]

'207.62 / 97.13'

In [154]:
gen_programs[79]

'207.62 - 97.13'

In [155]:
print(user_prompt)

### Context ###

shareholder return performance the line graph below compares the annual percentage change in ball corporation fffds cumulative total shareholder return on its common stock with the cumulative total return of the dow jones containers & packaging index and the s&p composite 500 stock index for the five-year period ended december 31 , 2012 .
it assumes $ 100 was invested on december 31 , 2007 , and that all dividends were reinvested .
the dow jones containers & packaging index total return has been weighted by market capitalization .
total return to stockholders ( assumes $ 100 investment on 12/31/07 ) total return analysis .

||12/31/2007|12/31/2008|12/31/2009|12/31/2010|12/31/2011|12/31/2012|
|---|---|---|---|---|---|---|
|Ball Corporation|$100.00|$93.28|$117.01|$155.14|$164.09|$207.62|
|DJ US Containers & Packaging|$100.00|$61.55|$84.76|$97.78|$96.27|$107.76|
|S&P 500|$100.00|$61.51|$75.94|$85.65|$85.65|$97.13|


source : bloomberg l.p .
aecharts .

### Question ###

t

||Amount (In Millions)|
|---|---|
|2016 net revenue|$705.4|
|Volume/weather|(18.2)|
|Retail electric price|13.5|
|Other|2.4|
|2017 net revenue|$703.1|

In [12]:
result = {
    "model": "gpt-3.5",
    "messages": [{"role": "system", "content": system_prompt}, 
                 {"role": "user", "content": user_example1}, 
                 {"role": "assistant", "content": assistant_example1}],
    "responses": responses,
}


with open("gpt-3.5.json", "w") as f:
    json.dump(result, f, indent=4) 

In [10]:
# num_cpus = 1

# ray.init(num_cpus=num_cpus)

# all_responses = []
# obj_ids = []

# for i in range(num_cpus):
#     chunk_data = data[i::num_cpus]
#     obj_id = get_responses.remote(i, chunk_data)
#     obj_ids.append(obj_id)

# for obj_id in tqdm.tqdm(obj_ids):
#     all_responses.extend(ray.get(obj_id))